File: merge_god/agents/callbacks.py
"""
Callback implementations for agent event handling.

This module provides callback implementations for different contexts
(logging, notifications, dashboard updates, etc.)
"""

import json
import time
from collections.abc import Callable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from .claude_agent import AgentAction, AgentCallbacks

# Constants for display and truncation
MAX_THINKING_PREVIEW_LENGTH = 50  # Max length before truncating thinking content
MAX_THINKING_LOG_LENGTH = 100  # Max length for logged thinking content


class PRProcessingCallbacks:
    """
    Callbacks for PR processing events.

    This integrates with pr-loop.py's logging and notification systems.
    """

    def __init__(
        self,
        pr_number: int,
        log_json: Callable[[str, dict[str, Any]], None],
        send_notification: Callable[[str, str | None, str, list[str] | None], bool] | None = None,
        max_retries: int = 3,
    ):
        self.pr_number = pr_number
        self.log_json = log_json
        self.send_notification = send_notification
        self.current_task = None
        self.current_task_start = None
        self.action_count = 0
        self.max_retries = max_retries
        self.retry_count = 0
        self.error_count = 0
        self.task_count = 0

    def on_thinking(self, content: str) -> None:
        """Agent is thinking/planning"""
        # Log abbreviated thinking to avoid spam
        if len(content) > MAX_THINKING_PREVIEW_LENGTH:
            self.log_json("agent_thinking", {
                "pr_number": self.pr_number,
                "task": self.current_task,
                "content": content[:MAX_THINKING_LOG_LENGTH] + "...",
                "full_length": len(content),
            })

    def on_action(self, action: AgentAction) -> None:
        """Agent is taking an action"""
        self.action_count += 1

        self.log_json("agent_action", {
            "pr_number": self.pr_number,
            "action_number": self.action_count,
            "action_type": action.type,
            "target": action.target,
            "status": action.status,
            "timestamp": action.timestamp.isoformat(),
        })

        # Send notification for important actions
        if action.type in ["git_commit", "gh_comment", "merge_pr"] and self.send_notification:
            self.send_notification(
                f"PR #{self.pr_number}: {action.type}",
                f"Agent performed {action.type} on {action.target}",
                "default",
                ["robot", "white_check_mark"] if action.status == "completed" else ["robot", "warning"],
            )

    def on_progress(self, current: int, total: int) -> None:
        """Progress update"""
        percentage = (current / total) * 100 if total > 0 else 0

        self.log_json("agent_progress", {
            "pr_number": self.pr_number,
            "current": current,
            "total": total,
            "percentage": round(percentage, 1),
        })

    def on_error(self, error: Exception) -> bool:
        """
        Error occurred - return True to continue/retry, False to abort.

        Implements retry logic for transient errors with exponential backoff.
        """
        self.error_count += 1
        error_type = type(error).__name__
        error_msg = str(error)

        # Check if error is transient (retryable)
        transient_errors = [
            "RateLimitError",
            "APIConnectionError",
            "APITimeoutError",
            "ServiceUnavailableError",
            "InternalServerError",
            "ConnectionError",
            "Timeout",
        ]

        is_transient = any(err in error_type for err in transient_errors)

        # Determine if we should retry
        should_retry = is_transient and self.retry_count < self.max_retries

        self.log_json("agent_error", {
            "pr_number": self.pr_number,
            "task": self.current_task,
            "error": error_msg,
            "error_type": error_type,
            "action_count": self.action_count,
            "error_count": self.error_count,
            "retry_count": self.retry_count,
            "is_transient": is_transient,
            "will_retry": should_retry,
        })

        if should_retry:
            self.retry_count += 1
            backoff_delay = min(2 ** self.retry_count, 32)  # Max 32 seconds

            self.log_json("agent_retry", {
                "pr_number": self.pr_number,
                "task": self.current_task,
                "retry_attempt": self.retry_count,
                "max_retries": self.max_retries,
                "backoff_seconds": backoff_delay,
            })

            if self.send_notification:
                self.send_notification(
                    f"PR #{self.pr_number}: Retrying after error",
                    f"Attempt {self.retry_count}/{self.max_retries}, waiting {backoff_delay}s",
                    "default",
                    ["warning", "arrows_counterclockwise"],
                )

            time.sleep(backoff_delay)
            return True  # Continue/retry

        # Permanent error or max retries exceeded
        self.log_json("agent_abort", {
            "pr_number": self.pr_number,
            "task": self.current_task,
            "reason": "max_retries_exceeded" if is_transient else "permanent_error",
            "total_errors": self.error_count,
            "total_retries": self.retry_count,
        })

        if self.send_notification:
            reason = f"Max retries ({self.max_retries}) exceeded" if is_transient else "Permanent error"
            self.send_notification(
                f"PR #{self.pr_number}: Agent Aborted",
                f"{reason}: {error_msg[:100]}",
                "high",
                ["x", "warning"],
            )

        return False  # Abort


class DashboardCallbacks:
    """
    Callbacks for dashboard monitoring.

    This integrates with dashboard.py's RepoMonitor to show real-time
    agent activity in the TUI.
    """

    def __init__(self, monitor):
        """
        Args:
            monitor: RepoMonitor instance from dashboard.py
        """
        self.monitor = monitor

    def on_thinking(self, content: str) -> None:
        """Update dashboard with agent thinking"""
        if self.monitor.current_agent_invocation:
            self.monitor.current_agent_invocation.thinking_content = content[:200]

        # Add to logs
        self.monitor.logs.append(f"🤖 Thinking: {content[:60]}...")

    def on_action(self, action: AgentAction) -> None:
        """Update dashboard with agent action"""
        if self.monitor.current_agent_invocation:
            self.monitor.current_agent_invocation.actions.append(action)

        # Add to logs with emoji based on action type
        action_emoji = {
            "read_file": "📖",
            "edit_file": "✏️",
            "run_tests": "🧪",
            "git_commit": "💾",
            "gh_comment": "💬",
        }.get(action.type, "⚙️")

        status_emoji = {
            "executing": "⏳",
            "completed": "✅",
            "failed": "❌",
        }.get(action.status, "❓")

        self.monitor.logs.append(
            f"{action_emoji} {status_emoji} {action.type}: {action.target[:40]}",
        )

    def on_progress(self, current: int, total: int) -> None:
        """Update dashboard with progress"""
        if self.monitor.current_agent_invocation:
            self.monitor.current_agent_invocation.progress = (current, total)

        percentage = (current / total) * 100 if total > 0 else 0
        self.monitor.current_action = f"Progress: {current}/{total} ({percentage:.0f}%)"

    def on_error(self, error: Exception) -> bool:
        """Handle error in dashboard"""
        self.monitor.logs.append(f"❌ Error: {str(error)[:60]}...")

        # Set error state
        self.monitor.status = "error"
        self.monitor.current_action = f"Error: {str(error)[:40]}..."

        # For interactive mode, could prompt user here
        # For now, abort
        return False


class CompositeCallbacks:
    """
    Composite callback that forwards to multiple callback implementations.

    Useful when you want both logging AND dashboard updates.
    """

    def __init__(self, *callbacks: AgentCallbacks):
        self.callbacks = callbacks

    def on_thinking(self, content: str) -> None:
        for callback in self.callbacks:
            callback.on_thinking(content)

    def on_action(self, action: AgentAction) -> None:
        for callback in self.callbacks:
            callback.on_action(action)

    def on_progress(self, current: int, total: int) -> None:
        for callback in self.callbacks:
            callback.on_progress(current, total)

    def on_error(self, error: Exception) -> bool:
        """Return True only if ALL callbacks say continue"""
        results = [callback.on_error(error) for callback in self.callbacks]
        return all(results)


class LoggingCallbacks:
    """
    Simple logging-only callbacks for testing and debugging.
    """

    def __init__(self, log_file: str | None = None):
        self.log_file = log_file
        self.events: list[dict[str, Any]] = []

    def _log(self, event_type: str, data: dict[str, Any]) -> None:
        """Log an event"""
        event = {
            "timestamp": datetime.now(UTC).isoformat(),
            "type": event_type,
            "data": data,
        }
        self.events.append(event)

        # Print to console
        print(f"[{event_type}] {data}")

        # Write to file if configured
        if self.log_file:
            with Path(self.log_file).open("a") as f:
                f.write(json.dumps(event) + "\n")

    def on_thinking(self, content: str) -> None:
        self._log("thinking", {"content": content[:100]})

    def on_action(self, action: AgentAction) -> None:
        self._log("action", {
            "type": action.type,
            "target": action.target,
            "status": action.status,
        })

    def on_progress(self, current: int, total: int) -> None:
        self._log("progress", {"current": current, "total": total})

    def on_error(self, error: Exception) -> bool:
        self._log("error", {"error": str(error), "type": type(error).__name__})
        return False

File: agents/callbacks.ts
/**
 * Callback implementations for agent event handling.
 *
 * Ported from agents/callbacks.py. Provides callback implementations for
 * different contexts (logging, notifications, dashboard updates, etc.).
 *
 * Note: claude_agent.ts is not yet ported, so the {@link AgentAction} and
 * {@link AgentCallbacks} types are defined here and re-exported. Once
 * claude_agent.ts is ported, those definitions should move there and this
 * module should import them instead.
 */

import { appendFileSync } from "node:fs";

/**
 * Represents a single action taken by the agent.
 *
 * Mirrors the consumed fields of the AgentAction dataclass from
 * claude_agent.py.
 */
export interface AgentAction {
  type: string;
  target: string;
  details: Record<string, unknown>;
  status: string;
  timestamp: Date;
  result?: Record<string, unknown> | null;
  error?: string | null;
}

/** Protocol for agent callbacks (mirrors the Python typing.Protocol). */
export interface AgentCallbacks {
  onThinking(content: string): void;
  onAction(action: AgentAction): void;
  onProgress(current: number, total: number): void;
  onError(error: Error): boolean;
}

/** Structured JSON logger callback signature. */
export type LogJsonFn = (event: string, data: Record<string, unknown>) => void;

/** Notification sender callback signature. */
export type SendNotificationFn = (
  title: string,
  body: string | null,
  channel: string,
  tags: string[] | null,
) => boolean;

/** A single action recorded by {@link LoggingCallbacks}. */
export interface LoggedEvent {
  timestamp: string;
  type: string;
  data: Record<string, unknown>;
}

/** Structural type of the dashboard monitor consumed by DashboardCallbacks. */
export interface DashboardMonitor {
  current_agent_invocation: {
    thinking_content?: string;
    actions: AgentAction[];
    progress: [number, number];
  } | null;
  logs: string[];
  current_action: string;
  status: string;
}

/** Synchronously block for `ms` milliseconds (mirrors Python time.sleep). */
function sleepSync(ms: number): void {
  const buf = new SharedArrayBuffer(4);
  const view = new Int32Array(buf);
  Atomics.wait(view, 0, 0, ms);
}

/**
 * Callbacks for PR processing events.
 *
 * Integrates with pr-loop's logging and notification systems.
 */
export class PRProcessingCallbacks implements AgentCallbacks {
  readonly pr_number: number;
  readonly log_json: LogJsonFn;
  readonly send_notification: SendNotificationFn | null;
  readonly max_retries: number;
  current_task: string | null;
  current_task_start: Date | null;
  action_count: number;
  retry_count: number;
  error_count: number;
  task_count: number;

  constructor(
    pr_number: number,
    log_json: LogJsonFn,
    send_notification: SendNotificationFn | null = null,
    max_retries: number = 3,
  ) {
    this.pr_number = pr_number;
    this.log_json = log_json;
    this.send_notification = send_notification;
    this.current_task = null;
    this.current_task_start = null;
    this.action_count = 0;
    this.max_retries = max_retries;
    this.retry_count = 0;
    this.error_count = 0;
    this.task_count = 0;
  }

  /** Called when the agent is thinking/planning. */
  onThinking(content: string): void {
    if (content.length > 50) {
      this.log_json("agent_thinking", {
        pr_number: this.pr_number,
        task: this.current_task,
        content: content.slice(0, 100) + "...",
        full_length: content.length,
      });
    }
  }

  /** Called when the agent takes an action. */
  onAction(action: AgentAction): void {
    this.action_count += 1;

    this.log_json("agent_action", {
      pr_number: this.pr_number,
      action_number: this.action_count,
      action_type: action.type,
      target: action.target,
      status: action.status,
      timestamp: action.timestamp.toISOString(),
    });

    if (action.type === "git_commit" || action.type === "gh_comment" || action.type === "merge_pr") {
      if (this.send_notification) {
        this.send_notification(
          `PR #${this.pr_number}: ${action.type}`,
          `Agent performed ${action.type} on ${action.target}`,
          "default",
          action.status === "completed" ? ["robot", "white_check_mark"] : ["robot", "warning"],
        );
      }
    }
  }

  /** Called with progress updates. */
  onProgress(current: number, total: number): void {
    const percentage = total > 0 ? (current / total) * 100 : 0;

    this.log_json("agent_progress", {
      pr_number: this.pr_number,
      current,
      total,
      percentage: Math.round(percentage * 10) / 10,
    });
  }

  /**
   * Called on error.
   *
   * Implements retry logic for transient errors with exponential backoff.
   * Returns true to continue/retry, false to abort.
   */
  onError(error: Error): boolean {
    this.error_count += 1;
    const errorType = error.constructor.name;
    const errorMsg = String(error);

    const transientErrors = [
      "RateLimitError",
      "APIConnectionError",
      "APITimeoutError",
      "ServiceUnavailableError",
      "InternalServerError",
      "ConnectionError",
      "Timeout",
    ];

    const isTransient = transientErrors.some((err) => errorType.includes(err));
    const shouldRetry = isTransient && this.retry_count < this.max_retries;

    this.log_json("agent_error", {
      pr_number: this.pr_number,
      task: this.current_task,
      error: errorMsg,
      error_type: errorType,
      action_count: this.action_count,
      error_count: this.error_count,
      retry_count: this.retry_count,
      is_transient: isTransient,
      will_retry: shouldRetry,
    });

    if (shouldRetry) {
      this.retry_count += 1;
      const backoffDelay = Math.min(2 ** this.retry_count, 32);

      this.log_json("agent_retry", {
        pr_number: this.pr_number,
        task: this.current_task,
        retry_attempt: this.retry_count,
        max_retries: this.max_retries,
        backoff_seconds: backoffDelay,
      });

      if (this.send_notification) {
        this.send_notification(
          `PR #${this.pr_number}: Retrying after error`,
          `Attempt ${this.retry_count}/${this.max_retries}, waiting ${backoffDelay}s`,
          "default",
          ["warning", "arrows_counterclockwise"],
        );
      }

      sleepSync(backoffDelay * 1000);
      return true;
    }

    this.log_json("agent_abort", {
      pr_number: this.pr_number,
      task: this.current_task,
      reason: isTransient ? "max_retries_exceeded" : "permanent_error",
      total_errors: this.error_count,
      total_retries: this.retry_count,
    });

    if (this.send_notification) {
      const reason = isTransient
        ? `Max retries (${this.max_retries}) exceeded`
        : "Permanent error";
      this.send_notification(
        `PR #${this.pr_number}: Agent Aborted`,
        `${reason}: ${errorMsg.slice(0, 100)}`,
        "high",
        ["x", "warning"],
      );
    }

    return false;
  }
}

/**
 * Callbacks for dashboard monitoring.
 *
 * Integrates with dashboard's RepoMonitor to show real-time agent activity in
 * the TUI.
 */
export class DashboardCallbacks implements AgentCallbacks {
  readonly monitor: DashboardMonitor;

  constructor(monitor: DashboardMonitor) {
    this.monitor = monitor;
  }

  /** Update dashboard with agent thinking. */
  onThinking(content: string): void {
    if (this.monitor.current_agent_invocation) {
      this.monitor.current_agent_invocation.thinking_content = content.slice(0, 200);
    }

    this.monitor.logs.push(`🤖 Thinking: ${content.slice(0, 60)}...`);
  }

  /** Update dashboard with agent action. */
  onAction(action: AgentAction): void {
    if (this.monitor.current_agent_invocation) {
      this.monitor.current_agent_invocation.actions.push(action);
    }

    const actionEmoji: Record<string, string> = {
      read_file: "📖",
      edit_file: "✏️",
      run_tests: "🧪",
      git_commit: "💾",
      gh_comment: "💬",
    };
    const statusEmoji: Record<string, string> = {
      executing: "⏳",
      completed: "✅",
      failed: "❌",
    };

    const ae = actionEmoji[action.type] ?? "⚙️";
    const se = statusEmoji[action.status] ?? "❓";

    this.monitor.logs.push(`${ae} ${se} ${action.type}: ${action.target.slice(0, 40)}`);
  }

  /** Update dashboard with progress. */
  onProgress(current: number, total: number): void {
    if (this.monitor.current_agent_invocation) {
      this.monitor.current_agent_invocation.progress = [current, total];
    }

    const percentage = total > 0 ? (current / total) * 100 : 0;
    this.monitor.current_action = `Progress: ${current}/${total} (${percentage.toFixed(0)}%)`;
  }

  /** Handle error in dashboard. Always aborts. */
  onError(error: Error): boolean {
    const msg = String(error);
    this.monitor.logs.push(`❌ Error: ${msg.slice(0, 60)}...`);

    this.monitor.status = "error";
    this.monitor.current_action = `Error: ${msg.slice(0, 40)}...`;

    return false;
  }
}

/**
 * Composite callback that forwards to multiple callback implementations.
 *
 * Useful when you want both logging AND dashboard updates.
 */
export class CompositeCallbacks implements AgentCallbacks {
  readonly callbacks: AgentCallbacks[];

  constructor(...callbacks: AgentCallbacks[]) {
    this.callbacks = callbacks;
  }

  onThinking(content: string): void {
    for (const callback of this.callbacks) {
      callback.onThinking(content);
    }
  }

  onAction(action: AgentAction): void {
    for (const callback of this.callbacks) {
      callback.onAction(action);
    }
  }

  onProgress(current: number, total: number): void {
    for (const callback of this.callbacks) {
      callback.onProgress(current, total);
    }
  }

  /** Return true only if ALL callbacks say continue. */
  onError(error: Error): boolean {
    const results = this.callbacks.map((callback) => callback.onError(error));
    return results.every((r) => r);
  }
}

/** Simple logging-only callbacks for testing and debugging. */
export class LoggingCallbacks implements AgentCallbacks {
  readonly log_file: string | null;
  readonly events: LoggedEvent[];

  constructor(log_file: string | null = null) {
    this.log_file = log_file;
    this.events = [];
  }

  /** Log an event. */
  private log(event_type: string, data: Record<string, unknown>): void {
    const event: LoggedEvent = {
      timestamp: new Date().toISOString(),
      type: event_type,
      data,
    };
    this.events.push(event);

    console.log(`[${event_type}] ${JSON.stringify(data)}`);

    if (this.log_file) {
      appendFileSync(this.log_file, JSON.stringify(event) + "\n", "utf8");
    }
  }

  onThinking(content: string): void {
    this.log("thinking", { content: content.slice(0, 100) });
  }

  onAction(action: AgentAction): void {
    this.log("action", {
      type: action.type,
      target: action.target,
      status: action.status,
    });
  }

  onProgress(current: number, total: number): void {
    this.log("progress", { current, total });
  }

  onError(error: Error): boolean {
    this.log("error", { error: String(error), type: error.constructor.name });
    return false;
  }
}
